nixos/modules/services/networking/headscale.nix

```diff
@@ -441,9 +448,6 @@ in {
           export HEADSCALE_DB_PASS="$(head -n1 ${escapeShellArg cfg.settings.db_password_file})"
         ''}
 
-        ${optionalString (cfg.settings.oidc.client_secret_file != null) ''
```

Comment:
`client_secret_file` is a Nix module specific option, the option in headscale's config is called `client_secret_path` so removing this will break peoples configuration and we would need to mark it as replaced.

Reply:
Honestly, I did not notice the option was renamed. I can mark `client_secret_file` as replaced, and nudge users to `client_secret_file`.

However, that way there will be "action by the distance": it will look like the user is configuring a headscale option; but that option will be replaced behind their back by NixOS module.

Thinking about it, I am more inclined to move it out of `settings`, so the users have some expectation that the option will be interpreted by Nix.

Reply:
OK, I tried to take a stab at this. Since all options are strictly specified, I think it makes sense to keep `client_secret_path` where it is, and rename `client_secret_file` to that setting. If I do such rename:

```
    (mkRenamedOptionModule ["services" "headscale" "settings" "oidc" "client_secret_file"] ["services" "headscale" "settings" "oidc" "client_secret_path"])
```

NixOS will complain:
```
error: evaluation aborted with the following error message: 'cannot find attribute `services.headscale.settings.oidc.client_secret_file''
```

That's probably because `settings` is already defined. If I remove the `settings`, it works as intended. Thus I am not sure if there is a way to change `client_secret_file` to `client_secret_path`. I grepped around in nixpks and did not find anything close.

Suggestions?

Reply:
Ok, I am not sure if this makes too much sense now that `client_secret_path` is in headscale.

We have a couple of options here (please tell me I am wrong if I am, this is still a bit confusing):

1. Have **two** options: `settings.oidc.client_secret_path` exposed straight into the config and `settings.oidc.client_secret_file`, which should be renamed out of settings to be NixOS _and_ systemd specific. If you give a path to `settings.oidc.client_secret_path` it will just put it in the config, and allow headscale to use it.

If you pass an option to the "new/renamed" option, it will read it into the credential store and give it to headscale via the new path. These options are mutual exclusive.

2. Remove `settings.oidc.client_secret_file` and this yanky secret reading and only let headscale handle `settings.oidc.client_secret_path` itself. No use of systemd cred store.

For simplicity, I think option 2 is the most appealing, allowing us to remove a NixOS specific config option and custom logic.

I am finding it hard to see the value of reading the configuration via the credential store, you will first read a unencrypted file off disk, and then put it into the systemd service runtime space. The unencrypted file will still be left on disk?

Currently the only advantage I see is that the original unencrypted file can potentially have root read only permission/very strict permissions? This might be good enough, but if there are not anything else, we need to check that this is the case at least.

Happy to be convinced, this was a bit to take in.

Reply:
I agree with the second approach. I will update the PR shortly.

> I am finding it hard to see the value of reading the configuration via the credential store, you will first read a unencrypted file off disk, and then put it into the systemd service runtime space. The unencrypted file will still be left on disk?

Systemd also supports reading encrypted credentials from disk. Which is just another argument to remove the systemd-specific changes from this PR.

> Currently the only advantage I see is that the original unencrypted file can potentially have root read only permission/very strict permissions? This might be good enough, but if there are not anything else, we need to check that this is the case at least.

That's exactly right. In my example all secrets can only read by root. Permissions are handled by systemd services via credentials system.